lib/connection/apm.js

```diff
@@ -214,23 +218,60 @@ class CommandFailedEvent {
    * @param {Object} command the command
    * @param {MongoError|Object} error the generated error or a server error response
    * @param {Array} started a high resolution tuple timestamp of when the command was first sent, to calculate duration
+   * @param {Array} metadata an array of command metadata
    */
-  constructor(pool, command, error, started) {
+  constructor(pool, command, error, started, metadata) {
     const cmd = extractCommand(command);
     const commandName = extractCommandName(cmd);
+    metadata = (Array.isArray(metadata) && metadata) || [];
+    const commandMetadata = new CommandMetadata(metadata);
 
     Object.assign(this, {
       duration: calculateDurationInMs(started),
       commandName,
       failure: maybeRedact(commandName, error),
       requestId: command.requestId,
-      connectionId: generateConnectionId(pool)
+      connectionId: generateConnectionId(pool),
+      metadata: commandMetadata
+    });
+  }
+}
+
+class BSONSerializationData {
+  /**
+   * Create serialize operation data, to store data about the serialize operation within a command's execution
+   *
+   * @param {Array} serializedBuffers the array of buffers that were serialized
+   * @param {Array} started a high resolution tuple timestamp of when the command started to serialize, to calculate duration
+   */
+  constructor(serializedBuffers, started) {
+    serializedBuffers = serializedBuffers || [];
+    started = started || [];
+    Object.assign(this, {
+      type: 'bsonSerialization',
+      buffersLength: serializedBuffers.reduce((s, b) => s + b.length, 0),
+      duration: calculateDurationInMs(started)
     });
   }
 }
 
+class CommandMetadata {
+  /**
+   * Create command metadata, to store data about operations that take place within a command's execution
+   *
+   * @param {Array} stepsData an array of command step data
+   */
+  constructor(stepsData) {
+    stepsData.reduce((self, step) => {
+      self[step.type] = step;
```

Comment:
would you also `delete step.type`? otherwise you end up with some duplication:
``` javascript
metadata: { 'bsonSerialization' : { type: 'bsonSerialization', ... } }
```